### ISSUE TITLE
Add Sequence builder rule when consuming another iterator

### ISSUE THREAD

**Issue by hfhbd** [detekt/detekt#5525]
## Expected Behavior of the rule
non-compliant:
```kotlin
fun <T> Sequence<T>.toString(): Sequence<String> {
  val iterator = iterator()
  return sequence {
    iterator.forEach {
      yield(it.toString())
    }
  }
}
```

compliant:
```kotlin
fun <T> Sequence<T>. toString(): Sequence<String> {
  return sequence {
    val iterator = iterator()
    iterator.forEach {
      yield(it.toString())
    }
  }
}
```

## Context
If you want to create your own sequence operators using `sequence {}`, you should support `constrainOnce()`. This implementation enforces that you do not store the iterator in a variable outside of the sequence builder. You must call it inside the scope.

**Comment by hfhbd** [comment]
Still relevant

**Comment by BraisGabin** [comment]
The example is valid but I'm not sure how this rule could be implemented. It's difficult to check correct behaviour on a static analysis.

I'd be glad to review a PR to fix something like this but the solution should be generic enought to merge it as a rule in the default set. But if someone wants to give it a try go for it. If, at the end, it doesn't make it inside the default set you can create your own custom rule and list it on our [marketplace](https://detekt.dev/marketplace)

**Comment by schalkms** [comment]
I agree with Brais here. This is a very specific rule. To implement this rule in a sophisticated manner, you'd need data flow analysis, which detekt does not have. Viewing the list of open issues, I don't think this rule would be a good candidate to dedicate effort to. Anyway, surely, we do accept contributions to the marketplace.
